scripting/include/srccoop/entitypatch.inc

```diff
@@ -1611,3 +1611,73 @@ public MRESReturn Hook_MusicTrackAceptInput(int _this, Handle hReturn, Handle hP
 	return MRES_Ignored;
 }
 
+//------------------------------------------------------
+// func_trackautochange
+// Prevent track from being blocked by players and forced to go back down
+//------------------------------------------------------
+public MRESReturn Hook_TrackChangeBlocked(int _this, Handle hReturn, Handle hParams)
+{
+	if (!DHookIsNullParam(hParams, 1))
+	{
+		CBasePlayer pPlayer = DHookGetParam(hParams, 1);
+		if (pPlayer.IsValid())
+		{
+			pPlayer.Suicide();
+			
+			DHookSetReturn(hReturn, false);
+			return MRES_Supercede;
+		}
+	}
+	
+	return MRES_Ignored;
+}
+
+//------------------------------------------------------
+// func_tracktrain
+// Prevent trains from being blocked by players
+// Moves players up if not blocked, otherwise kills blocking player
+//------------------------------------------------------
+public MRESReturn Hook_TrackTrainBlocked(int _this, Handle hReturn, Handle hParams)
+{
+	static float vecOriginAdjusted[3];
+	
+	if (!DHookIsNullParam(hParams, 1))
+	{
+		CBasePlayer pPlayer = DHookGetParam(hParams, 1);
+		if (pPlayer.IsValid())
+		{
+			pPlayer.GetAbsOrigin(vecOriginAdjusted);
+			vecOriginAdjusted[2] += 64.0;
+			if (TR_PointOutsideWorld(vecOriginAdjusted))
```

Comment:
I think the stock CanTeleportPlayer could be used here to check if origin is valid.

Reply:
I have changed it to use CanTeleportPlayer, and also added the `CAN_TP_MUSTDUCK` to StartDucking if needed.